crates/sui-types/src/supported_protocol_versions.rs

```diff
@@ -36,4 +41,61 @@ impl SupportedProtocolVersions {
     pub fn is_version_supported(&self, v: ProtocolVersion) -> bool {
         v.as_u64() >= self.min.as_u64() && v.as_u64() <= self.max.as_u64()
     }
+
+    pub fn as_range(&self) -> RangeInclusive<u64> {
+        self.min.as_u64()..=self.max.as_u64()
+    }
+
+    pub fn truncate_below(self, v: ProtocolVersion) -> Self {
+        let min = std::cmp::max(self.min, v);
+        Self { min, max: self.max }
+    }
+}
+
+/// Models the set of protocol versions supported by a validator.
+/// The `sui-node` binary will always use the SYSTEM_DEFAULT constant, but for testing we need
+/// to be able to inject arbitrary versions into SuiNode.
+#[derive(Serialize, Deserialize, Debug, Clone, Hash, PartialEq, Eq)]
+pub struct SupportedProtocolVersionsWithHashes {
+    pub versions: Vec<(ProtocolVersion, Digest)>,
+}
+
+impl SupportedProtocolVersionsWithHashes {
+    pub fn get_version_digest(&self, v: ProtocolVersion) -> Option<Digest> {
+        self.versions
+            .iter()
+            .find(|(version, _)| *version == v)
+            .map(|(_, digest)| *digest)
+    }
+
+    // Ideally this would be in sui-protocol-config, but sui-types depends on sui-protocol-config,
+    // so it would introduce a circular dependency.
+    fn protocol_config_digest(config: &ProtocolConfig) -> Digest {
+        let intent = Intent {
+            scope: IntentScope::ProtocolConfigDigest,
+            version: IntentVersion::V0,
+            app_id: AppId::Sui,
+        };
+        let mut digest = DefaultHash::default();
+        bcs::serialize_into(&mut digest, &intent).expect("serialization cannot fail");
```

Comment:
Could you explain why we need to include the intent in the digest computation? (and comments too) It doesn't seem necessary to me

Reply:
I suppose we don't, there's no need for domain separation here. Was just copying the typical pattern from elsewhere in the code.